File: barlus_gstreamer_proxy_original/barlus_gstreamer_proxy/gstreamer_proxy.py
# Copyright 2024, Evan Palmer
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

from typing import Any

import cv2 as cv
import gi
import gi.repository
import numpy as np
import rclpy
from cv_bridge import CvBridge
from rclpy.node import Node
from rclpy.qos import (
    DurabilityPolicy,
    HistoryPolicy,
    QoSProfile,
    ReliabilityPolicy,
    qos_profile_system_default,
)
from sensor_msgs.msg import CameraInfo, Image

gi.require_version("Gst", "1.0")
from gi.repository import Gst  # noqa: E402

from barlus_gstreamer_proxy.gstreamer_proxy_parameters import (  # noqa: E402
    gstreamer_proxy_parameters,
)


def gst_to_numpy(sample: Any) -> np.ndarray:
    """Convert a GStreamer sample to a numpy array."""
    buf = sample.get_buffer()
    caps = sample.get_caps()

    return np.ndarray(
        (
            caps.get_structure(0).get_value("height"),
            caps.get_structure(0).get_value("width"),
            3,
        ),
        buffer=buf.extract_dup(0, buf.get_size()),
        dtype=np.uint8,
    )


class GStreamerProxy(Node):
    def __init__(self) -> None:
        super().__init__("gstreamer_proxy")

        self.param_listener = gstreamer_proxy_parameters.ParamListener(self)
        self.params = self.param_listener.get_params()

        self.image_pub = self.create_publisher(
            Image, "/barlus/image_raw", qos_profile_system_default
        )
        self.camera_info_pub = self.create_publisher(
            CameraInfo,
            "/barlus/camera_info",
            QoSProfile(
                reliability=ReliabilityPolicy.RELIABLE,
                durability=DurabilityPolicy.TRANSIENT_LOCAL,
                history=HistoryPolicy.KEEP_LAST,
                depth=1,
            ),
        )

        camera_info = CameraInfo()
        camera_info.header.stamp = self.get_clock().now().to_msg()
        camera_info.header.frame_id = "/barlus/camera_link"
        camera_info.height = self.params.frame.height
        camera_info.width = self.params.frame.width
        camera_info.distortion_model = self.params.distortion_model
        camera_info.d = self.params.distortion_coefficients
        camera_info.k = self.params.camera_matrix
        camera_info.p = self.params.projection_matrix

        # This uses transient local durability, so the message will persist for future
        # subscribers
        self.camera_info_pub.publish(camera_info)

        # Configure the GStreamer pipeline
        self.bridge = CvBridge()
        self.video_pipe, self.video_sink = self.configure_stream(
            self.params.stream_address
        )

    def configure_stream(self, address: str) -> tuple[Any, Any]:
        Gst.init(None)

        video_source = f"rtspsrc location={address} latency=0 protocols=tcp"
        video_codec = "! rtph264depay ! h264parse ! avdec_h264"
        video_decode = "! decodebin ! videoconvert ! video/x-raw,format=(string)BGR"
        video_sink_conf = (
            "! queue ! appsink emit-signals=true sync=false max-buffers=1 drop=true"
        )

        command = " ".join([video_source, video_codec, video_decode, video_sink_conf])

        video_pipe = Gst.parse_launch(command)
        video_pipe.set_state(Gst.State.PLAYING)
        video_sink = video_pipe.get_by_name("appsink0")

        def proxy_frame_cb(sink: Any) -> gi.repository.Gst.FlowReturn:
            frame = gst_to_numpy(sink.emit("pull-sample"))
            grey = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
            self.get_logger().info(f"Frame shape: {grey.shape}")
            self.image_pub.publish(self.bridge.cv2_to_imgmsg(grey, encoding="mono8"))
            return gi.repository.Gst.FlowReturn.OK

        video_sink.connect("new-sample", proxy_frame_cb)

        return video_pipe, video_sink


def main(args=None):
    rclpy.init(args=args)
    proxy_node = GStreamerProxy()
    rclpy.spin(proxy_node)
    proxy_node.destroy_node()
    rclpy.shutdown()

File: barlus_gstreamer_proxy/src/gstreamer_proxy.cpp
#include "barlus_gstreamer_proxy/gstreamer_proxy.hpp"

namespace barlus
{

namespace
{

/// Convert a GStreamer sample to a cv::Mat image
auto gst_to_cv_mat(GstSample * sample) -> cv::Mat
{
  GstCaps * caps = gst_sample_get_caps(sample);
  GstBuffer * buffer = gst_sample_get_buffer(sample);
  GstStructure * structure = gst_caps_get_structure(caps, 0);

  int width;
  int height;
  gst_structure_get_int(structure, "width", &width);
  gst_structure_get_int(structure, "height", &height);

  GstMapInfo map;
  gst_buffer_map(buffer, &map, GST_MAP_READ);

  cv::Mat image(height, width, CV_8UC3, map.data);
  gst_buffer_unmap(buffer, &map);

  return image;
}

auto cv_mat_to_ros_image(const cv::Mat & image) -> sensor_msgs::msg::Image
{
  cv_bridge::CvImage cv_image;
  cv_image.image = image;
  cv_image.encoding = "bgr8";
  return *cv_image.toImageMsg();
}

}  // namespace

GStreamerProxy::GStreamerProxy(const rclcpp::NodeOptions & options)
: rclcpp_lifecycle::LifecycleNode("gstreamer_proxy", options)
{
}

auto GStreamerProxy::on_configure(const rclcpp_lifecycle::State & /*state*/) -> CallbackReturn
{
  try {
    param_listener_ = std::make_shared<gstreamer_proxy::ParamListener>(get_node_parameters_interface());
    params_ = param_listener_->get_params();
  }
  catch (const std::exception & e) {
    RCLCPP_ERROR(get_logger(), "Failed to get GStreamerProxy parameters: %s", e.what());  // NOLINT
    return CallbackReturn::ERROR;
  }

  image_pub_ = create_publisher<sensor_msgs::msg::Image>("/barlus/image_raw", rclcpp::SystemDefaultsQoS());
  camera_info_pub_ = create_publisher<sensor_msgs::msg::CameraInfo>("/barlus/camera_info", rclcpp::SystemDefaultsQoS());

  gst_init(nullptr, nullptr);

  // The following configurations are intended for low-latency transport
  const std::string video_source = "rtspsrc location=" + address + " latency=0 protocols=tcp";
  const std::string video_codec = "! rtph264depay ! h264parse ! avdec_h264";
  const std::string video_decode = "! decodebin ! videoconvert ! video/x-raw,format=(string)BGR";
  const std::string video_sink_conf = "! queue ! appsink emit-signals=true sync=false max-buffers=1 drop=true";
  const std::string command = video_source + video_codec + video_decode + video_sink_conf;

  GError * error = nullptr;
  pipeline_ = gst_parse_launch(command.c_str(), &error);

  if (error != nullptr) {
    RCLCPP_ERROR(get_logger(), "Failed to create GStreamer pipeline: %s", error->message);  // NOLINT
    g_error_free(error);
    return CallbackReturn::ERROR;
  }

  GstElement * sink = gst_bin_get_by_name(GST_BIN(pipeline_), "appsink0");

  gst_app_sink_set_emit_signals(GST_APP_SINK(sink), true);
  gst_app_sink_set_drop(GST_APP_SINK(sink), true);
  gst_app_sink_set_max_buffers(GST_APP_SINK(sink), 1);
  GstAppSinkCallbacks callbacks = {
    nullptr,
    [](GstAppSink * /*sink*/, gpointer /*user_data*/) -> GstFlowReturn { return GST_FLOW_OK; },  // Pre-roll
    [](GstAppSink * sink, gpointer /*user_data*/) -> void {                                      // Image received
      GstSample * sample = gst_app_sink_pull_sample(sink);
      const cv::Mat image = gst_to_cv_mat(sample);
      image_pub_->publish(cv_mat_to_ros_image(image));
    }};

  gst_element_set_state(pipeline_, GST_STATE_PLAYING);

  return CallbackReturn::SUCCESS;
}

auto GStreamerProxy::on_shutdown(const rclcpp_lifecycle::State & /*state*/) -> CallbackReturn
{
  gst_element_set_state(pipeline_, GST_STATE_NULL);
  gst_object_unref(pipeline_);

  return CallbackReturn::SUCCESS;
}

}  // namespace barlus
